### PR TITLE
[GAL-336] Adding system-paths setting to feature-pack

### PR DESCRIPTION
**Pull request by spyrkob**
Issue: https://issues.redhat.com/browse/GAL-336

related galleon-plugins PR: https://github.com/wildfly/galleon-plugins/pull/234

### CODE REVIEW

**Comment by spyrkob** [comment]
@jfdenise, I wasn't sure if the version of feature pack XSD needs to go up with this changes. If it does, let me know and I'll update the PR

**Comment by jfdenise** [comment]
@spyrkob , yes we should introduce a 3.0 xml and schema.

**Comment by spyrkob** [comment]
@aloubyansky / @jmesnil could you review this please?

**Comment by aloubyansky** [comment]
@spyrkob please update the docs to point people to the description of this feature.

**Comment by spyrkob** [comment]
@aloubyansky I expended the `Handling user-managed content` section to describe how the system-paths work and added example to `Feature-pack metadata (feature-pack.xml)`. Are there other sections that should be updated?

File: docs/guide/user-managed-content/index.adoc [hunk 1]
```diff
@@ -84,3 +84,73 @@ None
 
 |===
 
+### Protected system files
+
+An installation may contain files and directories users should not be changing - for example system libraries.
```
Comment (aloubyansky) [inline docs/guide/user-managed-content/index.adoc:89]:
```suggestion
An installation may contain files and directories that are meant to be managed exclusively by feature-packs and the provisioning tools and not by users. For example, core system libraries of the installation.
```

File: docs/guide/user-managed-content/index.adoc [hunk 2]
```diff
@@ -84,3 +84,73 @@ None
...
+Such paths can be marked as *system-paths* in the `feature-pack.xml`.
+
+When a file within a *system-path* is changed as part of an update, user changes are discarded and replaced with file from updated state. User changes are persisted in *.glold* files.
```
Comment (aloubyansky) [inline docs/guide/user-managed-content/index.adoc:92]:
```suggestion
If during an update a file matching a feature-pack's *system-path* was found to be modified by a user, the provisioning mechanism will create a copy of the user-modified file adding the *.glold* suffix to it and replace the original one with the file provided by the feature-pack.
```

File: docs/guide/user-managed-content/index.adoc [hunk 3]
```diff
@@ -84,3 +84,73 @@ None
...
+
+.User-added file within system-path
+[%header,cols="1,1a"]
+|===
+|Target in the updated state |Change applied to the updated state
```
Comment (aloubyansky) [inline docs/guide/user-managed-content/index.adoc:97]:
```suggestion
|Target path in the updated state |Change applied to the target path
```

File: docs/guide/user-managed-content/index.adoc [hunk 4]
```diff
@@ -84,3 +84,73 @@ None
...
+|Not found
+|The user-added file is copied to the updated state
+
+|Found and matches the file added by the user
+|
+None
+
+|Found and is different from the file added by the user
+|
+* A copy of the user-added file is created with suffix *.glold*
```
Comment (aloubyansky) [inline docs/guide/user-managed-content/index.adoc:108]:
```suggestion
* A copy of the user-added file is created with the *.glold* suffix
```

File: docs/guide/user-managed-content/index.adoc [hunk 5]
```diff
@@ -84,3 +84,73 @@ None
...
 |===
 
+### Protected system files
+
+An installation may contain files and directories users should not be changing - for example system libraries.
+Such paths can be marked as *system-paths* in the `feature-pack.xml`.
+
+When a file within a *system-path* is changed as part of an update, user changes are discarded and replaced with file from updated state. User changes are persisted in *.glold* files.
+
+.User-added file within system-path
```
Comment (aloubyansky) [inline docs/guide/user-managed-content/index.adoc:94]:
```suggestion
.User-added file matching a system-path
```

File: docs/guide/user-managed-content/index.adoc [hunk 6]
```diff
@@ -84,3 +84,73 @@ None
...
+None
+
+|Found and is different from the file added by the user
+|
+* A copy of the user-added file is created with suffix *.glold*
+
+* The file from updated state replaces the existing one
+|===
+
+.User-modified file within system-path
```
Comment (aloubyansky) [inline docs/guide/user-managed-content/index.adoc:113]:
```suggestion
.User-modified file matching a system-path
```

File: docs/guide/user-managed-content/index.adoc [hunk 7]
```diff
@@ -84,3 +84,73 @@ None
...
+[%header,cols="1,1a"]
+|===
+|Target in the updated state |Change applied to the updated state
```
Comment (aloubyansky) [inline docs/guide/user-managed-content/index.adoc:116]:
```suggestion
|Target path in the updated state |Change applied to the target path
```

File: docs/guide/user-managed-content/index.adoc [hunk 8]
```diff
@@ -84,3 +84,73 @@ None
...
+
+|Not found
+|The user-modified file is copied to the updated state and a warning is logged saying that the (equivalent of the) file modified in the previous state is not found in the updated state
```
Comment (aloubyansky) [inline docs/guide/user-managed-content/index.adoc:119]:
```suggestion
|The user-modified file is copied to the updated state and a warning is logged indicating that the (equivalent of the) file modified in the previous state is not found in the updated state
```

File: docs/guide/user-managed-content/index.adoc [hunk 9]
```diff
@@ -84,3 +84,73 @@ None
...
+
+|Found and matches the resulting file modified by the user
+|
+None
+
+|Found and is different from the resulting file modified by the user
+|
+* A copy of the user-added file is created with suffix *.glold*
```
Comment (aloubyansky) [inline docs/guide/user-managed-content/index.adoc:127]:
```suggestion
* A copy of the user-added file is created with the *.glold* suffix
```

File: docs/guide/user-managed-content/index.adoc [hunk 10]
```diff
@@ -84,3 +84,73 @@ None
...
+
+* The file from updated state replaces the existing one
```
Comment (aloubyansky) [inline docs/guide/user-managed-content/index.adoc:129]:
```suggestion
* The file from the updated state replaces the existing one
```

File: docs/guide/user-managed-content/index.adoc [hunk 11]
```diff
@@ -84,3 +84,73 @@ None
...
+
+|Found and matches the file added by the user
+|
+None
+
+|Found and is different from the file added by the user
+|
+* A copy of the user-added file is created with suffix *.glold*
+
+* The file from updated state replaces the existing one
```
Comment (aloubyansky) [inline docs/guide/user-managed-content/index.adoc:110]:
```suggestion
* The file from the updated state replaces the existing one
```

File: docs/guide/user-managed-content/index.adoc [hunk 12]
```diff
@@ -84,3 +84,73 @@ None
...
+None
+
+|Found and is different from the resulting file modified by the user
+|
+* A copy of the user-added file is created with suffix *.glold*
+
+* The file from updated state replaces the existing one
+|===
+
+.User-removed file or directory within system path
```
Comment (aloubyansky) [inline docs/guide/user-managed-content/index.adoc:132]:
```suggestion
.User-removed file or directory matching a system-path
```

File: docs/guide/user-managed-content/index.adoc [hunk 13]
```diff
@@ -84,3 +84,73 @@ None
...
+[%header,cols="1,1a"]
+|===
+|Target in the updated state |Change applied to the updated state
```
Comment (aloubyansky) [inline docs/guide/user-managed-content/index.adoc:135]:
```suggestion
|Target path in the updated state |Change applied to the target path
```

File: docs/guide/user-managed-content/index.adoc [hunk 14]
```diff
@@ -84,3 +84,73 @@ None
...
+
+|Not found
+|None
+
+|Found
+|The target is restored to the updated state.
+
+|===
+
+.User-created directory within system path
```
Comment (aloubyansky) [inline docs/guide/user-managed-content/index.adoc:145]:
```suggestion
.User-created directory matching a system-path
```

File: docs/guide/user-managed-content/index.adoc [hunk 15]
```diff
@@ -84,3 +84,73 @@ None
...
+[%header,cols="1,1a"]
+|===
+|Target in the updated state |Change applied to the updated state
```
Comment (aloubyansky) [inline docs/guide/user-managed-content/index.adoc:148]:
```suggestion
|Target path in the updated state |Change applied to the target path
```

File: core/src/main/java/org/jboss/galleon/layout/SystemPaths.java [hunk 16]
```diff
@@ -0,0 +1,53 @@
...
+import java.nio.file.Path;
+import java.util.HashSet;
+import java.util.Set;
+import java.util.stream.Collectors;
+
+public class SystemPaths {
+
+    private final Set<Path> paths;
+
+    public SystemPaths(Set<Path> paths) {
```
Comment (aloubyansky) [inline core/src/main/java/org/jboss/galleon/layout/SystemPaths.java:28]: Are these supposed to be always relative or always absolute? It'be good to make sure they always are one or the other.

Comment (spyrkob) [reply core/src/main/java/org/jboss/galleon/layout/SystemPaths.java:28]: Those should be relative paths only.
I added a check in `SystemPaths` and in `FeaturePackXmlParser30`

**Review by aloubyansky — APPROVED** [review]
Thanks @spyrkob!

**Pull request merged**